src/status_im/chat/models/images.cljs

```diff
@@ -104,30 +104,65 @@
 
 (re-frame/reg-fx
  ::camera-roll-get-photos
- (fn [[num end-cursor]]
+ (fn [[num end-cursor album]]
    (permissions/request-permissions
     {:permissions [:read-external-storage]
      :on-allowed  (fn []
-                    (-> (if end-cursor
-                            (.getPhotos
-                             CameraRoll
-                             #js
-                              {:first      num
-                               :after      end-cursor
-                               :assetType  "Photos"
-                               :groupTypes "All"
-                               :include    (clj->js ["imageSize"])})
-                            (.getPhotos CameraRoll
-                                        #js
-                                         {:first      num
-                                          :assetType  "Photos"
-                                          :groupTypes "All"
-                                          :include    (clj->js ["imageSize"])}))
+                    (-> (.getPhotos
+                         CameraRoll
+                         #js
+                          {:first      num
+                           :after      end-cursor
+                           :assetType  "Photos"
+                           :groupTypes (if (= album (i18n/label :t/recent)) "All" "Albums")
+                           :groupName  (when (not= album (i18n/label :t/recent)) album)
+                           :include    (clj->js ["imageSize"])})
                         (.then #(let [response (types/js->clj %)]
                                   (re-frame/dispatch [:on-camera-roll-get-photos (:edges response)
                                                       (:page_info response) end-cursor])))
                         (.catch #(log/warn "could not get camera roll photos"))))})))
 
+(re-frame/reg-fx
+ ::camera-roll-get-albums
+ (fn []
+   (let [albums (atom [{:title :smart-albums :data []}
+                       {:title (i18n/label :t/my-albums) :data []}])]
+     ;; Get the "recent" album first
+     (->
+       (.getPhotos CameraRoll
+                   #js
+                    {:first      1
+                     :groupTypes "All"})
+       (.then
+        (fn [res]
+          (let [recent-album {:title (i18n/label :t/recent)
+                              :count "--"
+                              :uri   (get-in (first (:edges (types/js->clj res))) [:node :image :uri])}]
+            (swap! albums update-in [0 :data] conj recent-album)
+            ;; Get albums, then loop over albums and get each one's cover (first photo)
+            (->
+              (.getAlbums CameraRoll #js {:assetType "All"})
+              (.then
+               #(let [response (types/js->clj %)]
+                  (reduce (fn [_ album]
```

Comment:
As we briefly talked about this problem @OmarBasem, we shouldn't use promises like this, inside a reduce function and without accumulating the results. I feel a huge urge to refactor this entire effect, it's far from ideal, but time is of essence :shrug: 

There are many things to consider here, but one is that any promise that fails inside `reduce` will disappear in a black hole, because there's no catch. The outer `catch` doesn't catch errors from the promises inside `reduce`.

The proper way is to **separate the side-effects from the pure part**. First build up a collection of promises to collect all photos for every album in the `CameraRoll`, then separately process them. There's no need to complect them.

Reply:
The tried your proposed solution with `Promise.all`, but the problem is that inside the `.then` call I wouldn't have access to the original albums sequence. Maybe I can get it using an atom wrapped around the whole method, but that wouldn't be too different, and also I would still have to loop again over the promises response to parse it. Maybe an async/await approach here would be better? https://github.com/status-im/status-mobile/pull/14867#discussion_r1086223887

Reply:
Given our priorities right now and the RC coming soon, I think the solution is good enough for now @OmarBasem. But this is the kind of problem that I think it's worth coming back later to refactor. Next week I'll probably find some time to attack this problem :space_invader:, unless you work yourself to refactor it of course. But hey, I don't mean to block the PR, sorry if that wasn't clear.

Reply:
Fine I will get back to it later. Thanks Icaro!